### ISSUE TITLE
Add initializedBy task dependency type (~see finalizedBy)

### ISSUE THREAD

**Issue by vlsi** [gradle/gradle#11945]
### Expected Behavior

```kotlin
val installPython by tasks.registering {
    // Install Python native executable
    // it will be triggered only in case some of the other tasks really requires python
}

val buildSomethingWithPython by tasks.registering {
    initializedBy(installPython)
    // use Python executable
}
```

`installPython` should not be executed if `buildSomethingWithPython` is up to date. For instance, if `buildSomethingWithPython` is taken from the build cache, then `installPython` should not be executed.

### Current Behavior

If `dependsOn(installPython)` is used, then `installPython` would be executed no matter if `buildSomethingWithPython` is up to date or not.

Currently, Gradle provides `finalizedBy` task relationship, however, it would be great to have `initializedBy` as well.

### Context

https://discuss.gradle.org/t/determine-if-a-task-is-up-to-date-from-cache-without-actually-executing-it/34504/5?u=vlsi

Note: I know [build services are coming](https://docs.gradle.org/nightly/userguide/build_services.html), however, build service does not provide a way to provide up-to-date checks for the service initialization.
On the other hand, tasks have understandable behavior when it comes to inputs, outputs, build cache, etc, etc, so it would be great (for the symmetry purpose :) ) to pair `finalizedBy` with `initializedBy`.

**Comment by chrisdennis** [comment]
The use case I have is the extraction of a kit bundle (zip) to a location under `$project.buildDir` so that tests can be executed against that kit. Logically this extraction task is an initializer for test execution and so `initializedBy` makes sense for me. I would also consider the outputs of my task to be logically 'part' of the `test` task as well though. When a user executes `cleanTest` I want the outputs of the `initializedBy` task to be deleted. Maybe this is something that should be true of both `initializedBy` and `finalizedBy`? Maybe this hints that there is a more formal concept of composite tasks (or equivalently sub-tasks) lurking here?

**Comment by nhoughto** [comment]
Yep this would be great for build dependency setup, i'd really like to not have to do `npm install` and unpack 200k tiny files, just for the task to be `up-to-date` or `from-cache`. Its a giant waste of time!

**Comment by nhoughto** [comment]
Another benefit of this would be that it would improve the value of the Gradle Scan / Enterprise features, at the moment build service provisioning time isn't split out, so the time taken to initialize/provision a Task and its BuildService (the current workaround for this problem) is included in the overall task execution time. My assumption is with this approach the initializedBy task would have its own task and execution time on the timeline and you would be able to see the real timing. 👍

**Comment by mlopatkin** [comment]
Sorry for the late reply!

This issue needs a decision from the team responsible for that area. They have been informed. Response time may vary.

**Comment by mlopatkin** [comment]
This feature request is in the backlog of the relevant team and is prioritized by them.

***

This is complicated because of the way up-to-date checks run. On a very high level, Gradle runs up-to-date checks at the moment it attempts to execute the task, so it is too late to schedule something _before_ that node.
We might be able to implement some other type of cacheable work, which will not be a task (like a FlowAction or a ValueSource - but these have to be made cacheable first).

**Comment by Vampire** [comment]
> Gradle runs up-to-date checks at the moment it attempts to execute the task, so it is too late to schedule something before that node.

Maybe something along the lines of:
- if `buildSomethingWithPython` is added to the task graph, also add `installPython`
- add an `onlyIf` to `installPython` that checks whether `buildSomethingWithPython` is up-to-date or not
?

Of course this adds race conditions like the up-to-dateness changing between this check and when the task is actually run. :-/

Maybe the proper solution would really be to do the work in a shared build service instead.
